core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+/**
+ * In an audit workflow, new data is written to an orphan snapshot that is not committed as the table's
+ * current state until it is audited. After auditing a change, it may need to be applied or cherry-picked
+ * on top of the latest snapshot instead of the one that was current when the audited changes were created.
+ *
+ * This class adds support for cherry-picking the changes from an orphan snapshot by applying them to
+ * the current snapshot. The output of the operation is a new snapshot with the changes from cherry-picked
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<AppendFiles> implements CherryPick {
```

Comment:
This should be `MergingSnapshotProducer<CherryPick>`. The type parameter is the type returned by common methods that configure the merging snapshot producer, like `set`. That way, when you call `set`  you get a `CherryPick` back and can call `fromSnapshotId`. Otherwise, you'd get an `AppendFiles` back (which is why you can't implement it right now).